src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs

```diff
@@ -122,155 +46,131 @@ private void OnSurfacePainted(object? sender, SKPaintSurfaceEventArgs e)
-				Cache.Remove(shadowsKey);
+				Cache.Remove(key);
 			}
 			else
 			{
-				canvas.DrawImage(shadowsImage, SKPoint.Empty);
+				canvas.DrawImage(snapshot, SKPoint.Empty);
 				canvas.Restore();
 				return;
 			}
 		}
 
-		float childWidth = (float)width * pixelRatio;
-		float childHeight = (float)height * pixelRatio;
-
-		float diffWidthSurfaceChild = surfaceWidth - childWidth;
-		float diffHeightSurfaceChild = surfaceHeight - childHeight;
+		var childWidth = (float)width * pixelRatio;
+		var childHeight = (float)height * pixelRatio;
+		var diffWidthSurfaceChild = surfaceWidth - childWidth;
+		var diffHeightSurfaceChild = surfaceHeight - childHeight;
 		canvas.Translate(diffWidthSurfaceChild / 2, diffHeightSurfaceChild / 2);
 
-		using var paint = new SKPaint();
-		paint.IsAntialias = true;
-
-		float cornerRadius = (float)_cornerRadius.BottomRight * pixelRatio;
+		using var paint = new SKPaint() { IsAntialias = true };
 
-		foreach (var shadow in shadows.Where(s => !s.IsInner))
+		foreach (var shadow in state.GetDropShadows())
 		{
-			var skShadow = SKShadow.From(shadow, childWidth, childHeight, cornerRadius, pixelRatio);
-
-			DrawDropShadow(canvas, paint, skShadow);
+			DrawDropShadow(state, canvas, paint, shadow);
 		}
 
-		// Always draw inner shadows on top of the drop shadows
-		if (hasInnerShadow)
-		{
-			var contentShape = new SKRoundRect(new SKRect(0, 0, childWidth, childHeight), cornerRadius);
-			canvas.ClipRoundRect(contentShape, antialias: true);
+		var contentShape = new SKRoundRect(
+			new SKRect(0, 0, childWidth, childHeight),
+			(float)state.CornerRadius.BottomRight * pixelRatio);
+		canvas.ClipRoundRect(contentShape, antialias: true);
 
-			// Draw the content background first
-			if (_currentContentBackgroundColor.HasValue)
-			{
-				var contentBackgroundColor = ToSkiaColor(_currentContentBackgroundColor.Value);
-				DrawContentBackground(canvas, contentBackgroundColor, contentShape);
-			}
-
-			// Then we draw the inner shadows
-			foreach (var shadow in shadows.Where(s => s.IsInner))
-			{
-				var skShadow = SKShadow.From(shadow, childWidth, childHeight, cornerRadius, pixelRatio);
+		if (background is { } bg)
+		{
+			DrawContentBackground(state, canvas, bg);
+		}
 
-				DrawInnerShadow(canvas, paint, skShadow);
-			}
+		foreach (var shadow in state.GetInsetShadows())
+		{
+			DrawInnerShadow(state, canvas, paint, shadow);
 		}
 
 		canvas.Restore();
 
-		if (!_shadowPropertyChanged)
+		// If a property has changed dynamically during this paint method,
+		// then we don't want to cache the updated shadows
+		if (!_isShadowDirty)
 		{
-			// If a property has changed dynamically, we don't want to cache the updated shadows
-			Cache.AddOrUpdate(shadowsKey, surface.Snapshot());
+			Cache.AddOrUpdate(key, surface.Snapshot());
 		}
-
-		_shadowPropertyChanged = false;
 	}
 
-	private bool ProcessContentBackgroundIfNeeded(out Color? contentBackgroundColor)
+	private static Color? GetBackgroundColor(Brush background)
```

Comment:
```suggestion
	private static Color? GetBackgroundColor(Brush? background)
```

Reply:
no need, the impl has built-in null handing logic that is needed.